lib/tpm2_session.c

```diff
@@ -164,3 +172,150 @@ tpm2_session *tpm2_session_new(TSS2_SYS_CONTEXT *sapi_context,
+ * a subtraction of 1 yields 0, which is a legal array size in C. In the false
+ * case (ie sizes not equal), 0 - 1 is -1, which will cause the compiler to
+ * complain.
+ */
+#define COMPILE_ASSERT_SIZE(a, b) \
+    typedef char WRONG_SIZE_##a[(sizeof(a) == sizeof(b)) - 1]
+
+// We check that the TSS library does not change sizes unbeknownst to us.
+COMPILE_ASSERT_SIZE(TPM2_HANDLE, UINT32);
+COMPILE_ASSERT_SIZE(TPMI_ALG_HASH, UINT16);
+COMPILE_ASSERT_SIZE(TPM2_SE, UINT8);
+
+tpm2_session *tpm2_session_restore(const char *path) {
+
+    tpm2_session *s = NULL;
+
+    FILE *f = fopen(path, "rb");
+    if (!f) {
+        LOG_ERR("Could not open path \"%s\", due to error: \"%s\"",
+                path, strerror(errno));
+        goto out;
+    }
+
+    uint32_t version;
+    bool result = files_read_header(f, &version);
+
+    TPM2_SE type;
+    result = files_read_bytes(f, &type, sizeof(type));
+    if (!result) {
+        LOG_ERR("Could not read session type");
+        goto out;
+    }
+
+    TPMI_ALG_HASH auth_hash;
+    result = files_read_16(f, &auth_hash);
+    if (!result) {
+        LOG_ERR("Could not read session digest algorithm");
+        goto out;
+    }
+
+    TPM2_HANDLE handle;
+    result = files_read_32(f, &handle);
+    if (!result) {
+        LOG_ERR("Could not read session handle");
+        goto out;
+    }
+
+    tpm2_session_data *d = tpm2_session_data_new(type);
+    if (!d) {
+        LOG_ERR("oom");
+        goto out;
+    }
+
+    tpm2_session_set_authhash(d, auth_hash);
+
+    s = tpm2_session_new(NULL, d);
+    s->output.session_handle = handle;
+
+out:
+    fclose(f);
+    return s;
+}
+
+bool tpm2_session_save(TSS2_SYS_CONTEXT *sapi_context, tpm2_session *session,
+        const char *path) {
+
+    bool result = false;
+
+    /*
+     * Perform a tpm context save/load but just write the
+     * context data in memory. This will mark the session
+     * with some RMs (like abrmd) as not-to-flush on client
+     * exit.
+     */
+    char *ptr;
+    size_t size = sizeof(TPMS_CONTEXT);
+    FILE *mem = open_memstream(&ptr, &size);
+    if (!mem) {
+        LOG_ERR("oom");
+        return false;
+    }
+
+    TPM2_HANDLE handle = tpm2_session_get_session_handle(session);
+    result = files_save_tpm_context_to_file(sapi_context, handle, mem);
+    if (!result) {
+        goto out;
+    }
+
+    rewind(mem);
+
+    TPM2_HANDLE dummy;
+    result = files_load_tpm_context_from_file(sapi_context, &dummy, mem);
+
+    FILE *session_file = fopen(path, "w+b");
+    if (!session_file) {
+        LOG_ERR("Could not open path \"%s\", due to error: \"%s\"",
+                path, strerror(errno));
+        goto out;
+    }
+
+    /*
+     * Now write the session_type, handle and auth hash data to disk
+     */
+    result = files_write_header(session_file, SESSION_VERSION);
+    if (!result) {
+         LOG_ERR("Could not write context file header");
+         goto out;
```

Comment:
You need a new error label here that also closes session_file before returning.

Reply:
Agreed.